### PR TITLE
Php phs 2

### PR DESCRIPTION
**Pull request by aaron-prindle**

### CODE REVIEW

**Comment by aaron-prindle** [comment]
when you composer require pkg version composer installs multiple files and updates the composer.json. it might be worth deleting all of the generated files except for what is is in /vendor

**Comment by aaron-prindle** [comment]
PTAL

File: ftl/php/builder.py [hunk 1]
```diff
@@ -33,27 +34,38 @@ def __init__(self, ctx, args, cache_version_str):
         super(PHP, self).__init__(ctx, _PHP_NAMESPACE, args, cache_version_str,
                                   [_COMPOSER_LOCK, _COMPOSER_JSON])
 
+    def _parse_composer_pkgs(self):
+        descriptor_contents = ftl_util.descriptor_parser(
+            self._descriptor_files, self._ctx)
+        composer_json = json.loads(descriptor_contents)
+        pkgs = []
+        for k, v in composer_json['require'].iteritems():
+            pkgs.append(k + ' ' + v)
```
Comment (dlorenc) [inline ftl/php/builder.py]: This feels a bit hacky - could we keep this as a tuple or some other structure? It looks like you split this back apart by space later on anyway.

**Comment by aaron-prindle** [comment]
getting timings locally:
full build of this composer.json:
https://gist.github.com/aaron-prindle/588998e1fe679af669cf5e92027b9a52

`2018-01-17,14:27:45.102 INFO     full build took 50 seconds`

rebuild with updated composer.json w/ single package:
https://gist.github.com/aaron-prindle/ef8ac052640eddca26fd1fc725b5734d

`2018-01-17,14:31:01.865 INFO     full build took 23 seconds`

File: ftl/php/builder.py [hunk 2]
```diff
@@ -33,27 +34,38 @@ def __init__(self, ctx, args, cache_version_str):
...
+            for pkg_txt in pkgs:
+                pkg = self.PackageLayer(self._ctx, self._descriptor_files,
+                                        pkg_txt, self._args.destination_path)
+                cached_pkg_img = None
                 if self._args.cache:
-                    with ftl_util.Timing("uploading pkg layer"):
-                        self._cash.Store(self._base, self._namespace,
-                                         pkg.GetCacheKey(), pkg.GetImage())
-            lyr_imgs.append(pkg)
+                    with ftl_util.Timing("checking cached pkg layer"):
```
Comment (dlorenc) [inline ftl/php/builder.py]: nit: i think you used underscores in here in other places?

File: ftl/php/builder.py [hunk 3]
```diff
@@ -93,8 +105,9 @@ def _gen_composer_install_tar(self, pkg_descriptor, destination_path):
             os.makedirs(app_dir)
 
             # Copy out the relevant package descriptors to a tempdir.
-            ftl_util.descriptor_copy(self._ctx, self._descriptor_files,
-                                     app_dir)
+            if pkg_descriptor is None:
```
Comment (dlorenc) [inline ftl/php/builder.py]: not sure I understand this part. Why do we only copy if it's none?

Comment (aaron-prindle) [reply ftl/php/builder.py]: if there is no pkg_descriptor, that means that composer is being run without a pkg in mind but with a composer.json file in its entirety.  When using the composer.json file, we need to have it copied.  When using a pkg_descriptor, we don't need to copy but can just use the string arg:
composer require <pkg> <version>
VS
composer install   # this uses composer.json in same dir

File: ftl/php/builder.py (hunk 2 above)
Comment (aaron-prindle) [reply ftl/php/builder.py]: i think I may have undid this in all places, so this is consistent currently.  Should I change all of these prints to be underscored?

Comment (dlorenc) [reply ftl/php/builder.py]: nah, it's fine as long as it's all consistent.

File: ftl/php/builder.py (hunk 1 above)
Comment (aaron-prindle) [reply ftl/php/builder.py]: I changed this to be a tuple

**Comment by aaron-prindle** [comment]
yes, the tests ensure that packages are still being installed correctly, the flags are still working properly, and images are still being uploaded.  The caching test was done separately to verify that this method works as expected, I can verify that only uncached packages are built on reupload

**Pull request merged**